rust/lance/src/dataset/fragment.rs

```diff
@@ -71,6 +72,28 @@ impl FileFragment {
         let reader = Box::new(reader);
         let (stream, schema) = reader_to_stream(reader).await?;
 
+        let schema = if matches!(params.mode, WriteMode::Append) {
+            // TODO: we need to pass down params somehow.
+            match DatasetBuilder::from_uri(dataset_uri).load().await {
+                Ok(dataset) => {
+                    schema.check_compatible(dataset.schema(), &Default::default())?;
+                    // Use the schema from the dataset, because it has the correct
+                    // field ids.
+                    dataset.schema().clone()
+                }
+                Err(Error::DatasetNotFound { .. }) => {
+                    // If the dataset does not exist, we can use the schema from
+                    // the reader.
+                    schema
+                }
```

Comment:
If the mode is append and the dataset does not exist isn't this an error?

Reply:
In `write_dataset` it's just a warning. It makes it a lot less complicated if we can make `append` default.